### PR TITLE
[Merged by Bors] - Support add and delete model dynamicly

### LINKED ISSUE #141: Support model update
Support update model (but not activate) via grpc/http interface

### PR DESCRIPTION
**Pull request by wangzhenyu32**
Closes #141

Signed-off-by: wangzhenyu32 <wang.zhenyu32@zte.com.cn>

### CODE REVIEW

File: adlik_serving/framework/manager/boarding_functor.cc [hunk 1]
```diff
@@ -285,8 +286,17 @@ struct ModelExecutor : ManagedNameVisitor {
...
-    if (auto action = next()) {
-      action->exec(store);
+    for (auto& name : names) {
+      actions.emplace_back(action(name));
+    }
+    while (true) {
+      auto action = next();
+      if (action) {
+        action->exec(store);
+        actions.erase(actions.begin());
```
Comment (amadeus-zte) [inline adlik_serving/framework/manager/boarding_functor.cc:296]: Be careful about Vector usage. Pls add some testcases for correct usage.

Comment (amadeus-zte) [inline adlik_serving/framework/manager/boarding_functor.cc:296]: Be careful about actions (Vector) usage. Should use origin pop (pop-front) or other direct api. Pls add some testcases for sure.

File: cub/env/fs/file_system.h [hunk 2]
```diff
@@ -35,6 +35,9 @@ DEFINE_ROLE(DirentVisitor) {
 DEFINE_ROLE(FileSystem) {
   ABSTRACT(ReadOnlyRegion * mmap(const std::string&) const);
   ABSTRACT(Status children(const std::string&, DirentVisitor&) const);
+  ABSTRACT(Status oneChildren(const std::string& dir, const std::string& childrenName, DirentVisitor& visitor) const);
```
Comment (amadeus-zte) [inline cub/env/fs/file_system.h:38]: oneChildren? should be oneChild

File: adlik_serving/framework/manager/boarding_functor.cc [hunk 3]
```diff
@@ -285,8 +286,17 @@ struct ModelExecutor : ManagedNameVisitor {
...
 
   void exec() {
-    if (auto action = next()) {
-      action->exec(store);
+    for (auto& name : names) {
+      actions.emplace_back(action(name));
+    }
+    while (true) {
+      auto action = next();
+      if (action) {
```
Comment (shiyingjin) [inline adlik_serving/framework/manager/boarding_functor.cc:294]: sort once(in select()) every loop??

File: cub/env/posix/posix_filesystem.cc [hunk 4]
```diff
@@ -115,8 +135,39 @@ Status PosixFileSystem::children(const std::string& dir, DirentVisitor& visitor)
...
+                                    DirentVisitor& visitor) const {
+  DirectReader reader(dir);
+  return reader.readOne(childrenName, visitor);
+}
+
 bool PosixFileSystem::exists(const std::string& fname) const {
   return access(fname.c_str(), F_OK) == 0;
 }
 
+Status PosixFileSystem::copyPath(const std::string& from, const std::string& to) const {
```
Comment (amadeus-zte) [inline cub/env/posix/posix_filesystem.cc:149]: copy a file from path... copyPath is not inappropriate function naming

File: adlik_serving/framework/manager/managed_store.cc [hunk 5]
```diff
@@ -94,6 +99,41 @@ cub::Status ManagedStore::start(const ModelId& id) {
...
+  if (ROLE(ModelStore).exist(modelName)) {
+    INFO_LOG << modelName << " model is already exist";
+    return cub::AlreadyExists;
+  }
+  std::string targetPath = cub::paths(ROLE(ModelOptions).getBasePath(), modelName);
+  cub::Status status = cub::filesystem().copyPath(path, targetPath);
+  if (cub::isFailStatus(status)) {
+    return status;
+  }
+  ROLE(ModelStore).configOneModel(modelName);
```
Comment (shiyingjin) [inline adlik_serving/framework/manager/managed_store.cc:113]: If there is a bad model, what are you going to do?

File: cub/env/posix/posix_filesystem.cc [hunk 6]
```diff
@@ -115,8 +135,39 @@ Status PosixFileSystem::children(const std::string& dir, DirentVisitor& visitor)
...
+  }
+  if (exists(to)) {
+    INFO_LOG << "target path of " << to << "is exist";
+    return cub::AlreadyExists;
+  }
+  sfs::copy(from, to, sfs::copy_options::recursive);
+  return cub::Success;
+}
+
+Status PosixFileSystem::deletePath(const std::string& fname) const {
```
Comment (amadeus-zte) [inline cub/env/posix/posix_filesystem.cc:163]: delete path , however input parameter is a filename. It's not clear.

File: adlik_serving/framework/manager/managed_store.cc [hunk 7]
```diff
@@ -94,6 +99,41 @@ cub::Status ManagedStore::start(const ModelId& id) {
...
+  ROLE(StorageLoop).once();
```
Comment (shiyingjin) [inline adlik_serving/framework/manager/managed_store.cc:114]: Add a model doesn't not need to load it

File: adlik_serving/apis/model_operate.proto [hunk 8]
```diff
@@ -0,0 +1,16 @@
+syntax = "proto3";
+
+package adlik.serving;
+option cc_enable_arenas = true;
+
+message ModelOperateRequest{
+  string model_name = 1;
```
Comment (shiyingjin) [inline adlik_serving/apis/model_operate.proto:7]: should think about model version

File: adlik_serving/framework/manager/managed_store.cc (hunk 7 above)
Comment (shiyingjin) [reply adlik_serving/framework/manager/managed_store.cc:114]: put these codes in model_operate_impl better. Let ManagedStore  not depend on StorageLoop, BoardingLoop, State Monitor better

File: adlik_serving/framework/manager/managed_store.h [hunk 9]
```diff
@@ -70,6 +78,10 @@ DEFINE_ROLE(ManagedStore) {
   USE_ROLE(ModelFactory);
   USE_ROLE(ModelStore);
   USE_ROLE(ServingStore);
+  USE_ROLE(ModelOptions);
+  USE_ROLE(StateMonitor);
+  USE_ROLE(StorageLoop);
+  USE_ROLE(BoardingLoop);
```
Comment (shiyingjin) [inline adlik_serving/framework/manager/managed_store.h:84]: too much coupling

File: adlik_serving/apis/model_operate_impl.cc [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+  std::string sourcePath = req.path();
+  INFO_LOG << "add model " << modelName;
+  tensorflow::Status status = tensorflow::Status::OK();
+  if (ROLE(ModelStore).exist(modelName)) {
+    INFO_LOG << modelName << " model is already exist";
+    FailureResponse(rsp, modelName + " model is already exist");
+    return status;
+  }
+  std::string targetPath = cub::paths(ROLE(ModelOptions).getBasePath(), modelName);
+  if (cub::isFailStatus(cub::filesystem().copyDir(sourcePath, targetPath))) {
```
Comment (shiyingjin) [inline adlik_serving/apis/model_operate_impl.cc:61]: Is sourcePath a directory or a zip file?

**Pull request closed without merging**